### PR TITLE
Validation

### PR DESCRIPTION
**Pull request by ondrejbouda**

### CODE REVIEW

File: src/StaticAnalysis/MessageHandlerValidationConfiguration.php [hunk 1]
```diff
@@ -0,0 +1,159 @@
...
+		$configuration->messageClassSuffix = 'Event';
+		$configuration->handlerClassSuffix = '';
+		$configuration->handlerClassPrefixRegex = '(.+)On';
+
+		return $configuration;
+	}
+
+
+
+	public function handlerClassMustBeFinal() : bool
```
Comment (xificurk) [inline src/StaticAnalysis/MessageHandlerValidationConfiguration.php]: No 'get' in the method names? That seems inconsistent with the rest of the library.

**Review by xificurk — CHANGES_REQUESTED** [review]
Making the `final` on message and handler optional opens a whole new can of worms and I'm not sure if that's wise...
Commands and/or handlers can now inherit from each other (which would probably break handler resolver logic). Also, they can be abstract and there is no validation for this.

File: src/StaticAnalysis/MessageHandlerValidationConfiguration.php (hunk 1 above)
Comment (ondrejbouda) [reply src/StaticAnalysis/MessageHandlerValidationConfiguration.php]: For now :) Ok, will correct.

**Comment by ondrejbouda** [comment]
In my previous commits I added support for custom validators.

File: src/StaticAnalysis/Rules/ClassNameHasSuffixRule.php [hunk 2]
```diff
@@ -0,0 +1,54 @@
...
+				$typeReflection->getName()
+			);
+		}
+	}
+
+
+
+	public function getRegexPattern() : string
+	{
+		return sprintf('#^(.*)%s$#', $this->suffix);
```
[the remark this answers is not in the document]
Comment (ondrejbouda) [reply src/StaticAnalysis/Rules/ClassNameHasSuffixRule.php]: Ir doesn't. There should be at least one character prepended. I changed the regex accordingly. This is the bare minimum.

File: src/StaticAnalysis/ValidateOnlyWhenTypeMatchesHandlerValidator.php [hunk 3]
```diff
@@ -0,0 +1,46 @@
...
+
+	public function __construct(HandlerType $type, IMessageHandlerValidator $validator)
+	{
+		$this->type = $type;
+		$this->validator = $validator;
+	}
+
+
+
+	public static function create(HandlerType $type, IMessageHandlerValidator $validator) : self
```
Comment (xificurk) [inline src/StaticAnalysis/ValidateOnlyWhenTypeMatchesHandlerValidator.php]: Unnecessary static factory - the signature is the same as public constructor

**Review by xificurk — CHANGES_REQUESTED** [review]

**Comment by ondrejbouda** [comment]
> Making the `final` on message and handler optional opens a whole new can of worms and I'm not sure if that's wise...
> Commands and/or handlers can now inherit from each other (which would probably break handler resolver logic). Also, they can be abstract and there is no validation for this.

This hole is already open. There is nothing that would make you to even use the validators in the first place. I suggest to let this be for now. The default is alright (strict pre-configured validators for commands and events), any other scenario is up to the user.

Maybe we could come up with a solution that somehow ties the validation logic and handler-resolving logic. But for now, I would leave it be.

**Review by xificurk — APPROVED** [review]

**Pull request merged**